### PR TITLE
Add VM validation webhook

### PR DESCRIPTION
**Pull request by rajivnathan**
Paired with https://github.com/codeready-toolchain/toolchain-e2e/pull/1070

### CODE REVIEW

**Review by alexeykazakov — APPROVED** [review]
Nice! Thanks.

**Comment by rajivnathan** [comment]
Removed the sandbox user check because in the tests the VM creation request has `"userInfo\":{\"username\":\"system:admin\"` and trying to unmarshall that to a User object causes an error. We don't really need it for now. If we want to enable setting `runStrategy` for some users we can handle it later.

File: deploy/webhook/member-operator-webhook.yaml [hunk 1]
```diff
@@ -275,6 +275,30 @@ objects:
       namespaceSelector:
         matchLabels:
           toolchain.dev.openshift.com/provider: codeready-toolchain
+    - name: users.virtualmachines.validating.webhook.sandbox
```
Comment (mfrancisc) [inline deploy/webhook/member-operator-webhook.yaml:278]: could you please add a comment explaining what the webhook does? as we have for the other ones above.

File: pkg/webhook/deploy/deployment_test.go [hunk 2]
```diff
@@ -227,7 +227,7 @@ func mutatingWebhookConfig(namespace, caBundle string) string {
...
...Bundle": "%[1]s","service": {"name": "member-operator-webhook","namespace": "%[2]s","path": "/validate-ssprequests","port": 443}},"failurePolicy": "Fail","matchPolicy": "Equivalent","name": "users.virtualmachines.ssp.webhook.sandbox","namespaceSelector": {"matchLabels": {"toolchain.dev.openshift.com/provider": "codeready-toolchain"}},"reinvocationPolicy": "Never","rules": [{"apiGroups": ["ssp.kubevirt.io"],"apiVersions": ["*"],"operations": ["CREATE","UPDATE"],"resources": ["ssps"],"scope": "Namespaced"}],"sideEffects": "None","timeoutSeconds": 5},{"admissionReviewVersions": ["v1"],"clientConfig": {"caBundle": "%[1]s","service": {"name": "member-operator-webhook","namespace": "%[2]s","path": "/validate-vmrequests","port": 443}},"failurePolicy": "Fail","matchPolicy": "Equivalent","name": "users.virtualmachines.validating.webhook.sandbox","namespaceSelector": {"matchLabels": {"toolchain.dev.openshift.com/provider": "codeready-toolchain"}},"reinvocationPolicy": "Never","rules": [{"apiGroups": ["kubevirt.io"],"apiVersions": ["*"],"operations": ["CREATE","UPDATE"],"resources": ["virtualmachines"],"scope": "Namespaced"}],"sideEffects": "None","timeoutSeconds": 5}]}`, caBundle, namespace)
```
Comment (mfrancisc) [inline pkg/webhook/deploy/deployment_test.go:230]: not related to this PR - but TBH I find this a bit difficult to maintain , maybe in the future we could remove/simplify this check/test.

File: pkg/webhook/validatingwebhook/validate_vm_request.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	hasRunStrategy, err := hasRunningStrategy(unstructuredRequestObj)
+	if err != nil {
+		log.Error(err, "failed to unmarshal VirtualMachine json object", "AdmissionReview", admReview)
+		return denyAdmissionRequest(admReview, errors.New("failed to validate VirtualMachine request"))
+	}
+	if hasRunStrategy {
+		log.Info("sandbox user is trying to create a VM with RunStrategy configured", "AdmissionReview", admReview) // not allowed because it interferes with the Dev Sandbox Idler
+		return denyAdmissionRequest(admReview, errors.New("this is a Dev Sandbox enforced restriction. Configuring RunStrategy is not allowed"))
+	}
+	// at this point, it is clear the user isn't a sandbox user, allow request
```
Comment (mfrancisc) [inline pkg/webhook/validatingwebhook/validate_vm_request.go:65]: minor - was this intentional ? maybe
```suggestion
	// the user is not creating a VM with the `runStrategy` configured, allowing the request.  
```

File: pkg/webhook/validatingwebhook/validate_vm_request_test.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	require.NoError(t, err)
+	body, err := io.ReadAll(resp.Body)
+	defer func() {
+		require.NoError(t, resp.Body.Close())
+	}()
+	require.NoError(t, err)
+	test.VerifyRequestBlocked(t, body, "this is a Dev Sandbox enforced restriction. Configuring RunStrategy is not allowed", "b6ae2ab4-782b-11ee-b962-0242ac120002")
+}
+
+func TestValidateVMAdmissionRequest(t *testing.T) {
```
Comment (mfrancisc) [inline pkg/webhook/validatingwebhook/validate_vm_request_test.go:39]: Not something to address now - but should we have all those tests going through the handler ? I mean merging this one with the `TestHandleValidateVMAdmissionRequestBlocked` into something like `TestHandleValidateVMAdmissionRequest`.

**Review by mfrancisc — COMMENTED** [review]
Great Job 👍 

I have only few minor comments/questions.

File: deploy/webhook/member-operator-webhook.yaml (hunk 1 above)
Comment (rajivnathan) [reply deploy/webhook/member-operator-webhook.yaml:284]: Good point, thanks, I missed that. Done c3a6cebff8433dc696d08f0dc82c7671e54c953a

File: pkg/webhook/validatingwebhook/validate_vm_request.go (hunk 3 above)
Comment (rajivnathan) [reply pkg/webhook/validatingwebhook/validate_vm_request.go:65]: Nope, not intentional. Thanks for fixing the comment 👍

File: pkg/webhook/validatingwebhook/validate_vm_request_test.go (hunk 4 above)
Comment (rajivnathan) [reply pkg/webhook/validatingwebhook/validate_vm_request_test.go]: You're right, I should have done it that way. Done 71f6795892683a2bf13289a9f90fae7d916b58eb

**Review by mfrancisc — APPROVED** [review]
Great Job 👍 🥇 

Thanks for addressing my comments 🙏

**Pull request merged**